File: src/Model/PSXMLReader.java
package edu.uwstout.berriers.PSHelper.Model;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Stack;
import java.util.zip.ZipEntry;
import java.util.zip.ZipException;
import java.util.zip.ZipFile;

import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamReader;

import org.apache.commons.io.FilenameUtils;

public class PSXMLReader {
	
	public static final XMLInputFactory XML_FACTORY = XMLInputFactory.newInstance();

	public static File checkForAndUpdatePath(XMLStreamReader reader, File currentFile) {

		// Is there a path tag we need to follow?
		if(reader.getAttributeValue(null, "path") != null) {

			// Retrieve relative path
			String newFilePath = reader.getAttributeValue(null, "path");

			// Fill in project name if needed
			if(newFilePath.contains("{projectname}")) {
				String baseFilename = FilenameUtils.getBaseName(currentFile.getPath());
				newFilePath = newFilePath.replaceAll("\\{projectname\\}", baseFilename);				
			}

			// Make into an absolute File Path object
			newFilePath = FilenameUtils.getFullPath(currentFile.getPath()) + newFilePath;			
			File newXMLFile = null;
			try { newXMLFile = new File(newFilePath).getCanonicalFile(); }
			catch(IOException e) {}
			
			return newXMLFile;
		}

		// Nothing to follow, so return the old path
		return currentFile;
	}
	
	// Examine the current reader tag for a 'path' attribute and return a new reader
	// to that file if one is found Otherwise, return the same reader.
	public static XMLStreamReader explodeTag(XMLStreamReader reader, Stack<File> currentFileStack)
			throws IOException, XMLStreamException {

		// Is there a path tag we need to follow?
		if(reader.getAttributeValue(null, "path") != null) {

			// Construct the absolute path to the new file and let it become the current file
			File newXMLFile = checkForAndUpdatePath(reader, currentFileStack.peek());
			currentFileStack.push(newXMLFile);

			// Turn into an input stream
			InputStream newXMLStream = getXMLStreamFromFile(newXMLFile);
			XMLStreamReader newReader = XML_FACTORY.createXMLStreamReader(newXMLStream);
			return newReader;
		}

		// Otherwise, stick with the passed in reader
		return reader;
	}
	
	@SuppressWarnings("resource")
	public static InputStream getXMLStreamFromFile(File pFile) throws IOException {

		// Sanity check
		if(pFile == null) { return null; }
	
		// The XML FileStream source
		InputStream lXMLFileStream = null;		
		String ext = FilenameUtils.getExtension(pFile.getName());
		
		switch(ext)
		{
			// Zip files with the XML inside them as doc.xml
			case "psz": case "zip":
			{
				// Declare variables that will be set inside of try-catch block
				ZipFile lZIPFile = null;
		
				// Try to open the PSZ file and obtain a stream for the doc.xml file
				try {
					lZIPFile = new ZipFile(pFile, ZipFile.OPEN_READ);
					ZipEntry lXMLFileEntry = lZIPFile.getEntry("doc.xml");
					if(lXMLFileEntry != null) {
						lXMLFileStream = lZIPFile.getInputStream(lXMLFileEntry);
					} else {
						throw new ZipException("could not find project XML file.");
					}
				} catch (Exception e) {
					if(lZIPFile != null) { lZIPFile.close(); }
					throw new IOException("Project File Error: general error while opening project file\n\t'" 
											+ pFile.getPath() + "'\n\t(" + e.getMessage() +").");
				}
			}
			break;
			
			// A raw PS xml file (probably to be accompanied by a .files directory)
			case "psx":
				lXMLFileStream = new FileInputStream(pFile);
			break;
		}		
		
		return lXMLFileStream;
	}
	
	// Function to read arrays encoded in XML (a la Agisoft's XML file format)
	public void readElementArray(XMLStreamReader reader, String arrayName, String elementName)
			throws XMLStreamException {
		
		// Loop over all elements till the end of the array is reached
		boolean readingElements = true;
		while(readingElements && reader.hasNext()) {

			// Move to the next tag element (START or END)
			reader.next();

			// Parse array element (preserving old reader, just in case)
			if(reader.isStartElement() && reader.getLocalName().equals(elementName))
			{
				XMLStreamReader oldReader = reader;
				processArrayElement(reader, elementName);
				reader = oldReader;
			}
			
			// Check for end of array
			else if(reader.isEndElement() && reader.getLocalName().equals(arrayName))
			{
				readingElements = false;
			}
		}
	}

	// Override this to process the specific array elements you expect to handle
	public void processArrayElement(XMLStreamReader reader, String elementName) {
		System.err.printf("Request to process array element '%s' not handled.\n", elementName);
	}
}

File: PSData/src/PSXMLReader.cpp
#include "PSXMLReader.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QXmlStreamReader>

#include <quazip/quazipfile.h>

// This looks for a 'path' attribute in the current element.
// Sometimes a portion of the XML file is stripped out and placed
// in it's own file under the .files directory. This function will
// look for that attribute and follow it if found.
QXmlStreamReader* PSXMLReader::explodeTag(QXmlStreamReader* reader, QStack<QFileInfo> currentFileStack) {
    // Is there a path tag we need to follow?
    if(!reader->attributes().value("", "path").isEmpty()) {
        // Construct the absolute path to the new QFile and let it become the current QFile
        QFileInfo newXMLFile = checkForAndUpdatePath(reader, currentFileStack.top());
        currentFileStack.push(newXMLFile);        

        // Turn into an input stream
        QXmlStreamReader* newReader = getXMLStreamFromFile(newXMLFile);
        return newReader;
    }

    // Otherwise, stick with the passed in reader
    return reader;
}

// This looks for a 'path' attribute in the current element and extracts
// a proper QFile string so it may be followed and opened.
QFileInfo PSXMLReader::checkForAndUpdatePath(QXmlStreamReader* reader, QFileInfo currentFile) {
    // Is there a path tag we need to follow?
    if(!reader->attributes().value("", "path").isEmpty()) {

        // Retrieve relative path
        QString newFilePath = reader->attributes().value("", "path").toString();

        qDebug("Opening %s with parent %s", newFilePath.toLocal8Bit().data(), currentFile.filePath().toLocal8Bit().data());

        // Fill in project name if needed
        if(newFilePath.contains("{projectname}")) {
            QString baseFilename = QFileInfo(currentFile).baseName();
            newFilePath = newFilePath.replace("{projectname}", baseFilename);
        }

        // Make into an absolute QFile Path object
        newFilePath = QFileInfo(currentFile.filePath()).absolutePath() + QDir::separator() + newFilePath;
        return QFileInfo(newFilePath);
    }

    // Nothing to follow, so file stays the same
    return currentFile;
}

// Examine the given PhotoScan file and create a reasonable XML stream reader for it.
// Will examine the file to find the XML data given:
// - A psz or zip file (will look for doc.xml inside the file)
// - OR a psx file accompanied by a .files directory
QXmlStreamReader* PSXMLReader::getXMLStreamFromFile(QFileInfo pFile) {

    // The XML FileStream source
    QXmlStreamReader* lXMLFileStream = NULL;
    QString ext = QFileInfo(pFile.fileName()).completeSuffix();

    // Zip files with the XML inside them as doc.xml
    if(ext == "psz" || ext == "zip")
    {
        QuaZipFile* lInsideFile = new QuaZipFile(pFile.filePath(), "doc.xml");
        if(!lInsideFile->open(QIODevice::ReadOnly)) {
            qWarning("Failed to open zip file: %d.", lInsideFile->getZipError());
            delete lInsideFile;
        } else if(lInsideFile->pos() >= 0) {
            lXMLFileStream = new QXmlStreamReader(lInsideFile);
        }
    } else if(ext == "psx" || ext == "xml") {
        // A raw PS xml QFile (probably to be accompanied by a .files directory)
        // Note: This also happens during testing
        QFile* lXMLFile = new QFile(pFile.absoluteFilePath());
        lXMLFile->open(QIODevice::ReadOnly);
        lXMLFileStream = new QXmlStreamReader(lXMLFile);
    } else {
        // Something else that's not supported
        qWarning("Unsupported file type pased to PSXMLReader: '%s'\n",
                 pFile.filePath().toLocal8Bit().data());
    }

    // Might be NULL
    return lXMLFileStream;
}

// Function to read arrays encoded in XML (a la Agisoft's XML File format)
void PSXMLReader::readElementArray(QXmlStreamReader* reader, QString arrayName, QString elementName) {
    // Loop over all elements till the end of the array is reached
    bool readingElements = true;
    while(readingElements && !reader->atEnd()) {
        // Move to the next tag element (START or END)
        reader->readNext();

        // Parse array element (preserving old reader, just in case)
        if(reader->isStartElement() && reader->name() == elementName) {
            QXmlStreamReader* oldReader = reader;
            processArrayElement(reader, elementName);
            reader = oldReader;
        }

        // Parse property tags inside of arrays (happens with depth maps)
        else if (reader->isStartElement() && reader->name() == "property") {
            QString lPropertyName = reader->attributes().value(NULL, "name").toString();
            QString lPropertyValue = reader->attributes().value(NULL, "value").toString();
            parseProperty(lPropertyName, lPropertyValue);
        }

        // Check for end of array
        else if(reader->isEndElement() && reader->name() == arrayName) {
            readingElements = false;
        }
    }
}

void PSXMLReader::parseProperty(const QString& pPropN, const QString& pPropV) {
    (void)pPropV;
    qWarning("Request to process property '%s' not handled.\n", pPropN.toLocal8Bit().data());
}

void PSXMLReader::processArrayElement(QXmlStreamReader* reader, QString elementName) {
    (void)reader;
    qWarning("Request to process array element '%s' not handled.\n", elementName.toLocal8Bit().data());
}
